cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -225,74 +291,126 @@ protected void run() {
     // Clear the interrupt flag
     Thread.interrupted();
 
-    if (stopped) {
+    if (killed) {
       // TODO: Kill the remote program execute on explicit stop on this monitor
     }
 
-    // Publish the remaining program state messages. We should retry indefinitely
-    String topicConfig = Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC;
-    String topic = requestKeyToLocalTopic.get(topicConfig);
-    try {
-      Retries.callWithRetries(() -> publish(topicConfig, topic, lastProgramStateMessages),
-                              RetryStrategies.fixDelay(1, TimeUnit.SECONDS), Retries.ALWAYS_TRUE);
-    } catch (Exception e) {
-      // This shouldn't happen since we have infinite retry.
-      LOG.error("Failed to publish remaining program state messages for program run {}", programRunId, e);
+    if (!lastProgramStateMessages.isEmpty()) {
+      String topicConfig = Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC;
+      String topic = requestKeyToLocalTopic.get(topicConfig);
+
+      // Publish last program state messages with Retries
+      Retries.runWithRetries(() -> Transactionals.execute(transactional, context -> {
+        AppMetadataStore store = AppMetadataStore.create(cConf, context, dsFramework);
+        publish(topicConfig, topic, lastProgramStateMessages, store);
+
+        // cleanup all the offsets after publishing terminal states.
+        for (String topicConf : requestKeyToLocalTopic.keySet()) {
+          store.deleteSubscriberState(topicConf, programRunId.getRun());
+        }
+      }), retryStrategy);
+    }
+  }
+
+  /**
+   * Detects if the remote runtime is in terminal state. If it is in terminal state, buffers all the program state 
+   * messages and updates the in memory store for next fetch
+   * 
+   * @param monitorResponses map of topic config to queue of monitor messages
+   */
+  private void updateProgramFinishTime(Map<String, Deque<MonitorMessage>> monitorResponses) {
+    if (monitorResponses.containsKey(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC)) {
+      Deque<MonitorMessage> monitorMessages = monitorResponses.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC);
+
+      if (programFinishTime < 0) {
+        programFinishTime = findProgramFinishTime(monitorMessages);
+      }
+
+      if (programFinishTime > 0) {
+        // Buffer the program state messages and don't publish them until the end
+        // Otherwise, once we publish, the deprovisioner will kick in and delete the cluster
+        // which could result in losing the last set of messages for some topics.
+        lastProgramStateMessages.addAll(monitorMessages);
```

Comment:
Better also remove the `PROGRAM_STATUS_EVENT_TOPIC` from the `monitorResponses`. This help simplifying the later logic in `processResponse` such that you don't need to check for `programFinishTime > 0` there.